biospecimen-tracking/api/src/main/java/org/phenotips/data/internal/controller/BiospecimensController.java

```diff
@@ -0,0 +1,379 @@
+            if (biospecimen == null) {
+                continue;
+            }
+            JSONObject obj = getJSONObjectFromBiospecimenData(biospecimen, propertiesToInclude, dateFormat);
+            if (obj != null) {
+                biospecimensJsonArray.put(obj);
+            }
+        }
+
+        if (biospecimensJsonArray.length() > 0) {
+            jsonObject.put(getName(), biospecimensJsonArray);
+        }
+    }
+
+    @Override
+    public PatientData<BiospecimenData> readJSON(JSONObject jsonObject)
+    {
+        if (jsonObject == null || jsonObject.optJSONArray(getName()) == null) {
+            return null;
+        }
+
+        JSONArray biospecimensJsonArray = jsonObject.getJSONArray(getName());
+
+        List<BiospecimenData> result = new LinkedList<>();
+
+        DateFormat jsonDateFormat =
+            new SimpleDateFormat(this.configurationManager.getActiveConfiguration().getISODateFormat());
+
+        try {
+            for (Object biospecimenObject : biospecimensJsonArray) {
+                BiospecimenData biospecimen = parseBiospecimenObject(biospecimenObject, jsonDateFormat);
+                if (biospecimen != null) {
+                    result.add(biospecimen);
+                }
+            }
+        } catch (ParseException e) {
+            this.logger.error("Unable to parse JSON data [{}]", e.getMessage());
+            return null;
+        }
+
+        return new IndexedPatientData<>(getName(), result);
+    }
+
+    /**
+     * Gets an EntityReference to the xWiki class representing BiospecimenData in the patient document.
+     * @return a EntityReference object
+     */
+    public static EntityReference getClassReference()
+    {
+        return new EntityReference(CLASS_NAME, EntityType.DOCUMENT, Constants.CODE_SPACE_REFERENCE);
+    }
+
+    /***************************************** Private Methods ****************************************/
+
+    /**
+     * Helper function. If selectedFieldNames is no null it will be returned, otherwise the PROPERTIES set will be
+     * returned otherwise.
+     *
+     * @param selectedFieldNames a collection of field names to return if not null
+     * @return a Collection
+     */
+    private static Collection<String> propertyNamesToIterateOver(Collection<String> selectedFieldNames)
+    {
+        if (CollectionUtils.isEmpty(selectedFieldNames)) {
+            return BiospecimenData.PROPERTIES;
+        } else {
+            return selectedFieldNames;
+        }
+    }
+
+    private static BiospecimenData parseBiospecimenObject(Object biospecimenObject, DateFormat jsonDateFormat)
+        throws ParseException {
+
+
+        if (!(biospecimenObject instanceof JSONObject)) {
+            return null;
+        }
+
+        JSONObject biospecimenJSONObject = (JSONObject) biospecimenObject;
+
+        BiospecimenData biospecimen = new BiospecimenData();
+
+        for (String property : BiospecimenData.PROPERTIES) {
```

Comment:
@danielpgross Yeah, that's because the code started with the parameters being a string map, and I let the date fields fall from one to the other and have only one implementation for them. I then changed the parameter data type and didn't remove the switch.